Implement Impersonation

ajibarra opened the pull request:
Refs #473

ajibarra committed:
wip Implement Impersonation #473

markstory commented on src/AuthenticationService.php:
```diff
@@ -428,4 +429,50 @@ public function getLoginRedirect(ServerRequestInterface $request): ?string
+     * @param ResponseInterface $response
+     * @return \ArrayAccess|array
+     */
+    public function stopImpersonate(ServerRequestInterface $request, ResponseInterface $response): \ArrayAccess|array
```
Should this method be `stopImpersonating()` or `stopImpersonation()`?

ajibarra replied:
Done

markstory commented on src/Authenticator/CookieAuthenticator.php:
```diff
@@ -252,4 +252,42 @@ protected function _createCookie($value): CookieInterface
+     */
+    public function isImpersonating(ServerRequestInterface $request, ResponseInterface $response): bool
+    {
+        // TODO: Implement isImpersonating() method.
```
Will you need the response for checking if impersonation is active? Seems like that could use only the request.

markstory commented on src/Authenticator/SessionAuthenticator.php:
```diff
@@ -115,4 +117,68 @@ public function clearIdentity(ServerRequestInterface $request, ResponseInterface
+        /** @var \Cake\Http\Session $session */
+        $session = $request->getAttribute('session');
+        if ($session->check($impersonateSessionKey)) {
+            //throw new UnauthorizedException('You are already impersonating a user.');
```
I think it reasonable to overwrite the existing impersonation if another is started.

ajibarra replied:
Makes sense

ajibarra replied:
Wait, doing some tests I found this use case:

* User A impersonates User B
* User B then impersonates User A
* User B (as User A) stops impersonation
* User B is logged in and no impersonation trace

So maybe we should not allow if user is impersonating already OR another way would be checking if user is being impersonated, and dont replace impersonator data but only impersonated one.

What do you think?

ADmad replied:
I don't think multi-level / nested impersonation should be allowed. It will make thinks overly complex.

markstory replied:
Agree that we should only support a single level of impersonation. The scenarios where I've used impersonation in the past are in a developer/support person needing to view a user's account as that user. I've not ever needed nested impersonation in that scenario.

markstory commented on src/Authenticator/SessionAuthenticator.php:
```diff
@@ -115,4 +117,68 @@ public function clearIdentity(ServerRequestInterface $request, ResponseInterface
+        $impersonateSessionKey = $this->getConfig('impersonateSessionKey');
+        /** @var \Cake\Http\Session $session */
+        $session = $request->getAttribute('session');
+        if (!$session->check($impersonateSessionKey)) {
+           // throw new UnauthorizedException('You are not impersonating a user.');
+        }
```
I think this is a reasonable error case to have.

ajibarra replied:
Removed //

markstory reviewed (commented):
You should run `vendor/bin/phpcbf` on your changes to fix the formatting issues.

ajibarra commented:
> You should run `vendor/bin/phpcbf` on your changes to fix the formatting issues.

Yeah..indeed it is still missing:

* ~Implement impersonation for CookieAuthenticator~ **NOT IMPLEMENTED (See #473)**
* ~Compatible with 7.X (Remove multiple return types, maybe something else)~
* ~Fix Code standards~
* ~Unit tests~
* ~Enable exceptions after finish implementation and tests~

But it is great to get some feedback about approach and code.

Thanks @markstory

markstory commented:
It would be helpful to know more about how you see developers using these features. Some use-cases that would be good to have examples of are:

1. How does a developer enable account impersonation?
2. Once active does impersonation stay active until it is stopped?
3. How does one turn off impersonation?

We'll need these use-cases to write the docs, so having them now will help when it is time to write the documentation. Other questions, I have around impersonation are:

* Should there be a time limit that impersonation is active for? If we don't want to build that into this plugin how would someone go about building that?
* Should we emit log messages when impersonation is activated, or should that be an application concern?

ajibarra committed:
Fix code standards. Adding errors back in SessionAuthenticate. Change methods to return request/response. Make it compatible with PHP 7.x

ajibarra committed:
Fix CookieAuthenticate::isImpersonating method signature

ajibarra committed:
Remove parameter comment

ajibarra committed:
Fix PHPstan issues

markstory commented on src/AuthenticationService.php:
```diff
@@ -215,6 +215,12 @@ public function clearIdentity(ServerRequestInterface $request, ResponseInterface
             if ($authenticator instanceof PersistenceInterface) {
+                if ($authenticator->isImpersonating($request)) {
+                    /** @psalm-var array{request: \Cake\Http\ServerRequest, response: \Cake\Http\Response} $stopImpersonationResult */
+                    $stopImpersonationResult = $authenticator->stopImpersonating($request, $response);
+                    $request = $stopImpersonationResult['request'];
+                    $response = $stopImpersonationResult['response'];
```
Should we use list destructuring instead? `[$request, $response] = $stopImpersonationResult;` ?

markstory commented on src/AuthenticationService.php:
```diff
@@ -428,4 +434,69 @@ public function getLoginRedirect(ServerRequestInterface $request): ?string
+        /** @var \Authentication\Authenticator\PersistenceInterface $provider */
+        $provider = $this->getAuthenticationProvider();
+        if (!($provider instanceof PersistenceInterface)) {
+            throw new \InvalidArgumentException(
+                'Provider must implement PersistenceInterface to be able to impersonate user.'
+            );
```
```suggestion
            $className = get_class($provider);
            throw new \InvalidArgumentException(
                "The {$className} Provider must implement PersistenceInterface in order to use impersonation.'
            );
```

markstory commented on src/AuthenticationService.php:
```diff
@@ -428,4 +434,69 @@ public function getLoginRedirect(ServerRequestInterface $request): ?string
+        /** @var \Authentication\Authenticator\PersistenceInterface $provider */
+        $provider = $this->getAuthenticationProvider();
+        if (!($provider instanceof PersistenceInterface)) {
+            throw new \InvalidArgumentException(
+                'Provider must implement PersistenceInterface to be able to impersonate user.'
+            );
```
```suggestion
            $className = get_class($provider);
            throw new \InvalidArgumentException(
                "The {$className} Provider must implement PersistenceInterface in order to use impersonation.'
            );
```

markstory commented on src/AuthenticationService.php:
```diff
@@ -428,4 +434,69 @@ public function getLoginRedirect(ServerRequestInterface $request): ?string
+        /** @var \Authentication\Authenticator\PersistenceInterface $provider */
+        $provider = $this->getAuthenticationProvider();
+        if (!($provider instanceof PersistenceInterface)) {
+            throw new \InvalidArgumentException(
+                'Provider must implement PersistenceInterface to be able to impersonate user.'
+            );
```
```suggestion
            $className = get_class($provider);
            throw new \InvalidArgumentException(
                "The {$className} Provider must implement PersistenceInterface in order to use impersonation.'
            );
```

markstory commented on src/Authenticator/PersistenceInterface.php:
```diff
@@ -41,4 +41,37 @@ public function persistIdentity(ServerRequestInterface $request, ResponseInterfa
+     * @param \Psr\Http\Message\ServerRequestInterface $request The request
+     * @return bool
+     */
+    public function isImpersonating(ServerRequestInterface $request): bool;
```
We will need a major release to extend the interfaces. We are likely going to have one of those soon those as there are a few other pull requests that require a major release.

ajibarra replied:
Since we are not going to implement impersonation for CookieAuthenticator, I have moved this code out to another interface ImpersonationInterface. I have added it to SessionAuthenticator only.

ajibarra replied:
I will be working on unit tests and please let me know if the target branch is ok or if we should make the MR to another branch. Thank you!

markstory replied:
This is fine. We can wait to do the major until this is merged. If folks urgently need changes they can use dev-master or pin the the sha they need.

markstory commented on src/Authenticator/SessionAuthenticator.php:
```diff
@@ -115,4 +117,80 @@ public function clearIdentity(ServerRequestInterface $request, ResponseInterface
+        $session = $request->getAttribute('session');
+        if ($session->check($impersonateSessionKey)) {
+            throw new UnauthorizedException(
+                'You are impersonating a user already. Stops impersonation before doing another one.'
```
```suggestion
                'You are impersonating a user already. Stop the current impersonation before impersonating another user.'
```

markstory commented on src/Authenticator/SessionAuthenticator.php:
```diff
@@ -115,4 +117,80 @@ public function clearIdentity(ServerRequestInterface $request, ResponseInterface
+        /** @var \Cake\Http\Session $session */
+        $session = $request->getAttribute('session');
+        if (!$session->check($impersonateSessionKey)) {
+            throw new UnauthorizedException('You are not impersonating a user.');
```
Should this be a silent error / treated as success instead? The developer wants impersonation to stop and if no impersonation is active then all impersonation is 'stopped'.

ajibarra committed:
Add new interface to move the code out from PersistenceInterface. Fix review comments and typos

ajibarra committed:
Removing methods from PersistenceInterface

ajibarra committed:
Adding ImpersonationInterface to AuthenticationService

ajibarra committed:
Fix list indexes

markstory commented on src/Authenticator/SessionAuthenticator.php:
```diff
@@ -115,4 +117,80 @@ public function clearIdentity(ServerRequestInterface $request, ResponseInterface
+        /** @var \Cake\Http\Session $session */
+        $session = $request->getAttribute('session');
+        if (!$session->check($impersonateSessionKey)) {
+            throw new UnauthorizedException('You are not impersonating a user.');
```
Should this be silent error? Won't raising an exception here cause the request to fail?

ajibarra replied:
it has been fixed in latest commit. I will mark it as resolved but feel free to reopen to discuss anything else about this.

```php
/** @var \Cake\Http\Session $session */
$session = $request->getAttribute('session');
if ($session->check($impersonateSessionKey)) {
    $identity = $session->read($impersonateSessionKey);
    $session->delete($impersonateSessionKey);
    $session->write($sessionKey, $identity);
    $this->setConfig('identify', true);
}
```

markstory reviewed (commented):
So how does one user an impersonated user once they have one? Are we going to support users 'being' the impersonated user and then 'reverting' their identity back to the old one? Will that be an `Authenticator` or are you thinking it should be a manual process?

markstory committed:
Fix deprecations in tests

* Update PHPUnit config for phpunit 9.
* Move to new fixture system.
* Don't mutate the new private properties on Uri.

markstory committed:
Update dev requirements

We'll need 4.4+ for tests now.

markstory committed:
Change cake dep

We need Uri class which was introducec in 4.3

markstory committed:
Require cake 4.4

I was going to try and work around the differences between 4.3 and 4.4
but fixtures cannot be upgraded.

Given that authentication is lined up to have a breaking change soon,
I don't feel so bad about this.

ajibarra commented:
> So how does one user an impersonated user once they have one? Are we going to support users 'being' the impersonated user and then 'reverting' their identity back to the old one? Will that be an `Authenticator` or are you thinking it should be a manual process?

If a user is already impersonating, then it throws an UnauthorizedException. We are going to support impersonation and also stops impersonation (not sure if this questions is related to multiple-levels impersonations. ATM we are going to support only single-level impersonation as previously discussed.

ajibarra commented:
> It would be helpful to know more about how you see developers using these features. Some use-cases that would be good to have examples of are:
> 
> 1. How does a developer enable account impersonation?
> 2. Once active does impersonation stay active until it is stopped?
> 3. How does one turn off impersonation?
> 
> We'll need these use-cases to write the docs, so having them now will help when it is time to write the documentation. Other questions, I have around impersonation are:
> 
> * Should there be a time limit that impersonation is active for? If we don't want to build that into this plugin how would someone go about building that?
> * Should we emit log messages when impersonation is activated, or should that be an application concern?

1. Impersonation is enabled by default and people just need to add a couple of actions like these to the UsersController (or similar):


```php
    public function impersonate($id = null)
    {
        $user = $this->Users->get($id);
        if ($this->Authentication->impersonate($user)) {
            $this->Flash->success(__('You are now impersonating {0}.', $user->email));
        } else {
            $this->Flash->error(__('The user could not be impersonated. Please, try again.'));
        }
        return $this->redirect(['action' => 'index']);
    }

    public function stopImpersonating()
    {
        $user = $this->Authentication->getIdentity();
        if ($this->Authentication->stopImpersonating()) {
            $this->Flash->success(__('You are not impersonating {0} anymore.', $user->email));
        } else {
            $this->Flash->error(__('An error has occurred leaving impersonation. Please, try again.'));
        }
        return $this->redirect(['action' => 'index']);
    }
```

2. Once active, impersonation stays active until session expires or user stops it.
3. To stop impersonation app must implement an action like previous one which calls AuthenticationComponent::stopImpersonating. If you mean turn off (disable feature) I think it is not needed since it depends exclusively from devs calling feature in their code. We are not providing a user the ability to impersonate if developer does not add it to the controller layer.

ajibarra committed:
Implement unit tests. Fix code standards.

ajibarra committed:
Merge branch 'fix-build' into issue/473

ajibarra committed:
Fix php version in ci

ajibarra committed:
Fix psalm issues

ajibarra commented:
@markstory I have merged your branch `fix-build` to be able to pass checks..Additionally I have fixed the php version missing in ci.yml

ajibarra committed:
Remove unused import

ajibarra committed:
Fix possible null value on impersonation

ajibarra committed:
Add more test cases for impersonation

ajibarra committed:
Fix remaining psalm coverage issues

ADmad commented on src/AuthenticationService.php:
```diff
@@ -214,10 +215,14 @@ public function authenticate(ServerRequestInterface $request): ResultInterface
     {
         foreach ($this->authenticators() as $authenticator) {
-            if ($authenticator instanceof PersistenceInterface) {
+            if ($authenticator instanceof ImpersonationInterface) {
```
Changing the interface check here means the identity won't be cleared for `CookieAuthenticator` which only implements `PersistenceInterface`, not `ImpersonationInterface`. I don't think that's expected.

ADmad commented on src/AuthenticationService.php:
```diff
@@ -31,7 +32,7 @@
  * Authentication Service
  */
-class AuthenticationService implements AuthenticationServiceInterface
+class AuthenticationService implements AuthenticationServiceInterface, ImpersonationInterface
```
Implementing `ImpersonationInterface` is not necessary since `AuthenticationServiceInterface` already extends it.

ADmad commented on src/AuthenticationService.php:
```diff
@@ -428,4 +433,72 @@ public function getLoginRedirect(ServerRequestInterface $request): ?string
+    public function isImpersonating(ServerRequestInterface $request): bool
+    {
+        /** @var \Authentication\Authenticator\ImpersonationInterface $provider */
+        $provider = $this->getAuthenticationProvider();
```
The code on 494 - 500 is repeated thrice. They could be moved into a separate protected method `getImpersonationProvider()`.

ajibarra replied:
Fixed

ADmad commented on src/AuthenticationServiceInterface.php:
```diff
@@ -17,12 +17,12 @@
 use Psr\Http\Message\ServerRequestInterface;
 
-interface AuthenticationServiceInterface extends PersistenceInterface
+interface AuthenticationServiceInterface extends ImpersonationInterface
```
If `AuthenticationServiceInterface` extends `ImpersonationInterface` then perhaps it should be under the namespace `Authentication` instead of `Authentication\Authenticator`.

Also changing the interface is not backwards compatible and would require a major version bump. To avoid that this extension would have to be keep and have `AuthenticationService` directly implement `ImpersonationInterface` (which is already done currently).

markstory replied:
:+1: to adding a new interface instead of modifying the existing one.

ajibarra replied:
I have reverted extends and created a method to check in AuthenticationComponent for the AuthenticationService implementation of ImpersonationInterface. Added a test for it as well.

ajibarra committed:
Fix review comments (clear identity, implements)

ajibarra committed:
Revert extends to ImpersonationInterface, change AuthComponent to get right Service

ajibarra committed:
Add test for getImpersonationAuthenticationServiceFailure

markstory reviewed (commented):
Looking pretty good to me. I'll write some docs later this week.
